### ISSUE TITLE
Using a negative (cyclic) indexation for Array, List

### ISSUE THREAD

**Issue by laugof** [dotnet/csharplang#1910]
-1 for the last element, -2 for the previous...

**Comment by jnm2** [comment]
This would be a breaking change, changing the meaning and consistency of existing programs.

**Comment by ufcpp** [comment]
```cs
using System;

public static class CyclicArray
{
    internal static int Mod(int i, int length)
    {
        i %= length;
        if (i < 0) i += length;
        return i;
    }

    public static CyclicArray<T> Cyclic<T>(this T[] array) => new CyclicArray<T>(array);
}

public struct CyclicArray<T>
{
    private readonly T[] _array;
    public CyclicArray(T[] array) => _array = array;
    public ref T this[int i] => ref _array[CyclicArray.Mod(i, _array.Length)];
}

class Program
{
    static void Main()
    {
        var data = new[] { 1, 2, 3, 4 }.Cyclic();

        for (int i = -10; i <= 10; i++)
        {
            Console.WriteLine(data[i]);
        }
    }
}
```

**Comment by CyrusNajmabadi** [comment]
This is not a language change.  If you'd like the BCL to behave this way, i would recommend bringing this up at dotnet/corefx.

**Comment by YairHalberstadt** [comment]
Closing as the equivalent has been implemented for C# 8.0 using the `^1` syntax. See https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-8#indices-and-ranges.

If you still want negative indices to work, you'll have to open an issue on https://github.com/dotnet/runtime
